kura/org.eclipse.kura.core.configuration/src/main/java/org/eclipse/kura/core/configuration/CloudConfigurationHandler.java

```diff
@@ -413,27 +415,71 @@ private void doExecSnapshot(CloudletTopic reqTopic, KuraPayload reqPayload, Kura
         respPayload.setBody(body);
     }
 
-    private static byte[] toResponseBody(Object o) throws KuraException {
+    private byte[] toResponseBody(Object o) throws KuraException {
         //
         // marshall the response
-        StringWriter sw = new StringWriter();
+        String result = null;
         try {
-            XmlUtil.marshal(o, sw);
+            result = marshalXml(o);
         } catch (Exception e) {
-            s_logger.error("Error marshalling snapshots: {}", e);
+            logger.error("Error marshalling snapshots: {}", e);
             throw new KuraException(KuraErrorCode.CONFIGURATION_SNAPSHOT_LOADING, e);
         }
 
         byte[] body = null;
         try {
-            body = sw.toString().getBytes("UTF-8");
+            body = result.getBytes("UTF-8");
         } catch (UnsupportedEncodingException e) {
-            s_logger.error("Error encoding response body: {}", e);
+            logger.error("Error encoding response body: {}", e);
             throw new KuraException(KuraErrorCode.CONFIGURATION_SNAPSHOT_LOADING, e);
         }
 
         return body;
     }
+
+    private ServiceReference<Marshalling>[] getXmlMarshallers() {
+        String filterString = String.format("(&(kura.service.pid=%s))", "org.eclipse.kura.marshalling.xml.provider");
+        return ServiceUtil.getServiceReferences(this.bundleContext, Marshalling.class, filterString);
+    }
+
+    private void ungetMarshallersServiceReferences(final ServiceReference<Marshalling>[] refs) {
+        ServiceUtil.ungetServiceReferences(this.bundleContext, refs);
+    }
+
+    protected XmlComponentConfigurations unmarshalXml(String xmlString) throws KuraException {
+        XmlComponentConfigurations result = null;
+        ServiceReference<Marshalling>[] marshallerSRs = getXmlMarshallers();
+        try {
+            for (final ServiceReference<Marshalling> marshallerSR : marshallerSRs) {
```

Comment:
Maybe break on success?

Reply:
I thought that too, but Sonar was complaining. I will review